### ISSUE TITLE
Credentials file created on ARM cannot be opened on Intel

### ISSUE THREAD

**Issue by unikitty37** [rails/rails#46137]
Credentials files created on an ARM-based machine cannot be opened on an Intel-based machine, and vice-versa. As ARM is big-endian and Intel is little-endian, I suspect the encryption might be affected by the byte order.

This is problematic, as I'm developing on an M1 Mac, but the CI and production are running on Intel CPUs. I can work around this by pulling an Intel Docker image and editing credentials in that, but that relies Docker continuing to support running Intel containers on ARM (it's also a major pain for people without Docker-based workflows :)

### Steps to reproduce

(I can't work out how to fit this test case into the executable test case templates — sorry!)

On an ARM-based machine:

```bash
% rails new testapp
% cd testapp
% rails db:create
% rails db:migrate
% rails credentials:edit
```

Save the credentials file, verify that `config/master.key` is present, and copy the `testapp` folder onto an Intel-based machine. On the Intel machine:

```bash
% cd testapp
% rails assets:precompile
````

### Expected behaviour

It precompiles the assets.

### Actual behaviour

It fails with the following message:

```
rails aborted!
ActiveSupport::MessageEncryptor::InvalidMessage: ActiveSupport::MessageEncryptor::InvalidMessage
…/testapp/config/environment.rb:7:in `<main>'
Caused by:
OpenSSL::Cipher::CipherError: 
…/testapp/config/environment.rb:7:in `<main>'
```

### System configuration
**Rails version**: Rails 7.0.3.1

**Ruby version**: ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [arm64-darwin21]

**Comment by byroot** [comment]
> As ARM is big-endian and Intel is little-endian

They're both little-endian AFAIK, so unlikely to be the cause.

**Comment by unikitty37** [comment]
Ah, yes; I misread the documentation. While ARM *can* technically support either, Apple's M1 (which the file was created on) is little-endian, so it's definitely not down to endianness.

To further confuse matters, I seem to be able to read the credentials on M1 when the file was created on Intel…

**Comment by byroot** [comment]
@unikitty37 could you:
  -  generate a test key
  - use it to generate a message on both M1 and Intel
  - Share them here.

Bonus point if you can make an executable test case using https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/guides/bug_report_templates/generic_main.rb

**Comment by Flixt** [comment]
TL;DR; I was not able to reproduce the issue on my Intel and M1 (ARM) machines.

---

I tried to reproduce this issue on the following machines and failed:

- Machine 1: MacBookPro 2021 (Apple M1 Max / ARM)
- Machine 2: MacBookPro 2019 (Intel Core i9)

Running this executable test succeeds on both machines. I'm not sure the test is correct though. The `key` and `encrypted_data` were generated on the M1.

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
end

require "active_support"
require "active_support/core_ext/object/blank"
require "minitest/autorun"

class BugTest < Minitest::Test
  def test_encrypt_decrypt
    crypt = ActiveSupport::MessageEncryptor.new(key)
    decrypted_message = crypt.decrypt_and_verify(encrypted_data)

    assert_equal 'my secret data', decrypted_message
  end

  def test_key
    assert_equal key, ActiveSupport::KeyGenerator.new('password').generate_key('salt', 32)
  end

  private
    def key
      # Same as ActiveSupport::KeyGenerator.new('password').generate_key('salt', 32)
      Base64.decode64("GzRd1V9io1rs25IpvHrpWzBajVOJQBNGJ+RvgtOkHl4=\n")
    end

    def encrypted_data
      # Generated by ActiveSupport::MessageEncryptor.new(key).encrypt_and_sign('my secret data')
      Base64.decode64("U2k5alluQklRVzlJU21GSE9IWktUMEZoVUdrMVpGZHZNR1VyT0ZCQ1ZtVjZU\naTlMVTBocE1HcHJXVDB0TFVsR1ptbFVaa0ZMVTBSNk1tVmpWblIwVm1Gb2NH\nYzlQUT09LS0wZTg4OTNhZmU4Nzc3N2YwMjc4YTc2ZmE4YWExYzVkM2Y5M2Zh\nM2Ri\n")
    end
end
```

Additionally I tried the steps described by @unikitty37 and created a fresh Rails project on the M1, changed the credentials on the M1 and tried to precompile the assets on the Intel. It worked fine. Even without the `master.key` being present. 

With the correct `master.key` file on the Intel machine I'm able to edit the credentials previously changed on the M1 machine.

What happened to me some time ago was, my IDE added a newline to the `config/credentials.yml.enc` file, because it is configured to add trailing new lines to every file, but this leads to another error when trying to edit the credentials then (`Couldn't decrypt config/credentials.yml.enc. Perhaps you passed the wrong key?`)

@unikitty37 If you want, clone my test app from https://github.com/Flixt/issue46137 and check if you are able to edit the credentials on your Intel based machine.
